tools/opam-updates.py

```diff
@@ -0,0 +1,326 @@
+        otherwise current path.
+        """
+        if self.new is not None:
+            return self.new_path(prefix)
+        return self.current_path(prefix)
+
+
+def download(session, url, file_path):
+    "Download the file at [url] in [file_path]."
+    r = session.get(url)
+    r.raise_for_status()
+    with open(file_path, 'wb') as f:
+        f.write(r.content)
+        f.flush()
+
+
+def update_files(session, auth, package, prefix):
+    """
+    Lookup the opam files for the [package], and update them in the
+    appropriate subfolder in [prefix].
+    """
+    base_url = package.github_api_url
+
+    r = session.get(base_url, auth=auth)
+    r.raise_for_status()
+    entries = r.json()
+
+    for entry in entries:
+        if entry["type"] == "dir":
+            continue
+        new_path = os.path.join(package.path(prefix), entry["name"])
+        download(session, entry["download_url"], new_path)
+
+    if any(entry["type"] == "dir"
+           and entry["name"] == "files"
+           for entry in entries):
+        r = session.get(os.path.join(base_url, "files"), auth=auth)
+        r.raise_for_status()
+        entries = r.json()
+        prefix = os.path.join(package.path(prefix), "files")
+        os.makedirs(prefix, exist_ok=True)
+
+        for entry in entries:
+            if entry["type"] == "dir":
+                continue
+            new_path = os.path.join(package.path(prefix), entry["name"])
+            download(session, entry["download_url"], new_path)
+
+
+def extract_upgrade(row):
+    "Extract upgraded packages from a row in the opam output."
+    name, cur, _, new = row.strip().split()[2:6]
+    return Package(name, cur, new)
+
+
+def extract_update(row):
+    "Extract updated packages from a row in the opam output."
+    name, ver = row.strip().split()[2:4]
+    return Package(name, cur=ver)
+
+
+def extract_install(row):
+    "Extract newly installed packages from a row in the opam output."
+    name, ver = row.strip().split()[2:4]
+    return Package(name, new=ver)
+
+
+def print_update_actions(upgrades, updates, new_packages):
+    """
+    Action pretty printer.
+    """
+    if upgrades:
+        print("Upgrades:")
+        for pkg in upgrades:
+            print("\t- {}: {}->{}".format(pkg.name, pkg.cur, pkg.new))
+        print()
+
+    if updates:
+        print("Updates (changed upstream):")
+        for pkg in updates:
+            print("\t- {} {}".format(pkg.name, pkg.cur))
+        print()
+
+    if new_packages:
+        print("New dependencies:")
+        for pkg in new_packages:
+            print("\t- {} {}".format(pkg.name, pkg.new))
+        print()
+
+
+def get_opam_upgrade_output():
+    """
+    Returns the output of `opam upgrade` (executed in dry mode) after
+    adding the upstream opam remote to the remotes.
+    """
+    call("opam remote add upstream https://opam.ocaml.org".split(' '))
+
+    try:
+        opam_out = check_output("opam upgrade --dry-run -y".split(' '))
```

Comment:
There is the `--json=file` option to dump the output in a nice json format into a file, maybe that will be easier to parse and less likely to break in the future?
```
opam upgrade --dry-run -y --json=/tmp/j
```

Reply:
It was my first attempt, unfortunately it turned out that the json is completely useless. It will list all the installed packages, without any metadata to know if they were updates, upgrades, installs or reinstalls.

Reply:
Oh ok  - I guess the only indicator that it's a reinstall is that the version field is the same as the current one.

Reply:
No, it could be an update (opam/url changed upstream but version kept unmodified) or a reinstall.

Reply:
It is really useless